pkg/controller/acmechallenges/update_test.go

```diff
@@ -0,0 +1,184 @@
+/*
+Copyright 2022 The cert-manager Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package acmechallenges
+
+import (
+	"context"
+	"errors"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	clienttesting "k8s.io/client-go/testing"
+	featuretesting "k8s.io/component-base/featuregate/testing"
+
+	"github.com/cert-manager/cert-manager/internal/controller/feature"
+	"github.com/cert-manager/cert-manager/pkg/client/clientset/versioned/fake"
+	utilfeature "github.com/cert-manager/cert-manager/pkg/util/feature"
+	"github.com/cert-manager/cert-manager/test/unit/gen"
+)
+
+func TestUpdateObjectStandard(t *testing.T) {
+	runUpdateObjectTests(t)
+}
+
+func TestUpdateObjectSSA(t *testing.T) {
+	t.Skip("Server Side Apply cannot be tested because PatchType is not supported by the fake versioned client")
+	defer featuretesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, feature.ServerSideApply, true)()
+	runUpdateObjectTests(t)
+}
+
+func runUpdateObjectTests(t *testing.T) {
+	simulatedUpdateError := errors.New("simulated-update-error")
+	simulatedUpdateStatusError := errors.New("simulated-update-status-error")
+
+	tests := []struct {
+		name         string
+		mods         []gen.ChallengeModifier
+		notFound     bool
+		apiResponse  clienttesting.ReactionFunc
+		panicMessage string
+		errorMessage string
+	}{
+		// Modifying the finalizers and any status fields results in both
+		// finalizers and status being updated.
+		{
+			name: "success",
+			mods: []gen.ChallengeModifier{
+				gen.SetChallengeFinalizers([]string{"example.com/another-finalizer"}),
+				gen.SetChallengePresented(true),
+			},
+		},
+		// If the API server responds with a NOT FOUND error, the error is
```

Comment:
These comments are great and help to understand what is the expected behaviour the function 👍🏼